### DISCUSSION TITLE
For optional types, undefined is more appropriate than null using typescript.

### DISCUSSION THREAD

**Discussion by ismail-codar** [odata2ts/odata2ts#221] [Q&A] [answered]
Model generator generates code like this:
``` ts
  /**
   *
   * OData Attributes:
   * |Attribute Mail| Attribute Value |
   * | --- | ---|
   * | Name | `Mail` |
   * | Type | `Edm.String` |
   */
  Mail: string | null;
  ```
But in this case TS2322 compile error occurs like the picture below.  It to be fix updating this line null to undefined . 
https://github.com/odata2ts/odata2ts/blob/main/packages/odata2ts/src/generator/ModelGenerator.ts#L144

Generating code as 
`Mail: string | undefined;` or `Mail?: string;` is more appropriate.

----

![image](https://github.com/odata2ts/odata2ts/assets/282104/b5c96291-d918-4ba5-9965-9261a8e10389)

**Comment by texttechne** [comment]
Hi @ismail-codar,

I absolutely agree with your statement and I would always prefer `undefined` over `null`.

However, the typing is based on the OData spec. Whenever you do a query on an entity / entitySet (without any $select) you get back all the fields which are then either filled or `null`. 

I'm not sure what the use case in your code snippet is, but when you want to create/update data, then you should use the `Editiable` model versions. In these versions stuff is either set or `undefined`...

Does this help?

**Comment by ismail-codar** [comment]
Thanks for your response Hubert. I would like to talk about an example case.
https://stackblitz.com/edit/stackblitz-starters-zp6nb2?file=src%2FApp.tsx (please see line 49)

In practice, as developers, we use various libraries. For example zod validation inference returns undefined for optional types. So this occurs typescript compile error on line 49.

Mostly undefined is used for optinal types in typescript.  Using null can sometimes be theoretically correct but not developer friendly. 

Can we provide a generator option for generating null or undefined values on models?

**Comment by texttechne** [comment]
Hi @ismail-codar, 

thanks for explaining your use case! So you're validating the input according to the generated model to submit this stuff back to your OData service.

In these kind of situations you should use the editable model versions. For example, when generating the legendary TrippinService we have a `Person` entity type. For this `odata2ts` will generate one model representing query results - by default called just `Person` and this is the one you're using and talking about and which uses `null` instead of `undefined`. But `odata2ts` will also generate one model version which should be used in your use case, i.e. the `EditablePerson`. The editable version will use `undefined` instead of `null` for optional attibutes and will leave out server managed fields (actually that's configurable, see [manages fields](https://odata2ts.github.io/docs/generator/configuration#managed-properties)) like ID, createdAt, ...

So in your case you should use the editble model versions, they're tailored for exactly that. To be honest, it was quite a suprise on my part to realize this distinction and get it sorted out... 

So did you try to use the editable model versions? Do you have any issues using them for your scenario?

**Comment by texttechne** [comment]
Hi @ismail-codar,

converted this into a discussion as it feels more appropriate here. Hope that's fine with you.

**Reply by ismail-codar** [reply]
Hi Hubert. I'm sorry for returning late. 
If we go through EditablePerson, the optional name field appears as a `string | null | undefined` and TS2322 error occours again. 

I'd like to set up a Trippin-based real example for us to think about. But in the coming days.

**Comment by texttechne** [comment]
Hi @ismail-codar,

ahh ok, yes of course: It can still be `null` and the lib you're using is still complaining. 

You could use a helper type in combination with the `EditableModel`:
```ts
type WithoutNull<T> = {
  [prop in keyof T]: Exclude<T[prop], null>;
};
```

My test: 
![image](https://github.com/odata2ts/odata2ts/assets/1963160/6940b4d2-ce9c-4362-923c-c59a3b71492f)

**Comment by texttechne** [comment]
Ok, and here some black magic for using the regular model instead of the EditableModel:

```ts
type IntersectionWithoutNull<A, B> = A & B extends infer U ? { [P in keyof U]: Exclude<U[P], null> } : never;

type Nullable<T> = {
  [Prop in keyof T]: Extract<T[Prop], null> extends null ? Prop : never;
}[keyof T];

type NonNullable<T> = {
  [Prop in keyof T]: Extract<T[Prop], null> extends null ? never : Prop;
}[keyof T];

type NullToUndefined<T> = IntersectionWithoutNull<Partial<Pick<T, Nullable<T>>>, Pick<T, NonNullable<T>>>;


// Testing
export interface TestModel {
  name: string;
  optional: string | null;
}

const x2: NullToUndefined<TestModel> = {
  name: "testing",
};

const y2: NullToUndefined<TestModel> = {
  name: "testing",
  optional: null, // should raise an error
};

const z2: NullToUndefined<TestModel> = {
  name: "testing",
  optional: "ho",
};
```

**Reply by ismail-codar** [reply]
I encountered a similar situation when using next-auth too. Everyone can not be to change their types. Type conversion is required; I am sorry. I am using the following method.
```ts
type NonNullableObject<T> = {
    [K in keyof T]-?: NonNullableObject<NonNullable<T[K]>>;
};
```

**Reply by texttechne** [reply]
Hi @ismail-codar,
this doesn't seem to be the correct solution as it removes `undefined` as well.
I've thought the task was to remove only `null` from the typing...

However, to finish off this discussion, please mark something as answer. Whatever seems most appropriate to you...

**Comment by ismail-codar** [comment]
Hi following code is not removing undefined but removes null.
```typescript
type NonNullableObject<T> = {
  [K in keyof T]?: NonNullableObject<NonNullable<T[K]>>;
};
```

**Reply by texttechne** [reply]
Hi, 
yes, this removes null, but makes ALL properties optional.

```ts
type NonNullableObject2<T> = {
  [K in keyof T]?: NonNullableObject2<NonNullable<T[K]>>;
};

interface EditableModel {
    id: string;
    opt?: string | null;
}

type Test2 = NonNullableObject2<EditableModel>;

// @ts-expect-error: id should be required
const w2: Test2 = {
}

const x2 : Test2 = {
    id: "hey"
}

const y2 : Test2 = {
    id: "hey",
    opt: "ho"
}
```

Case w2 won't throw an error anymore, because the required `id` property become optional.

**Reply by texttechne** [reply]
I think my first solution is the best choice, here it is slightly modified to account for nested structures:
```ts
type WithoutNull<T> = {
  [prop in keyof T]: WithoutNull<Exclude<T[prop], null>>;
};
```

This would work then without any errors:
```ts
interface EditableModel {
    id: string;
    opt?: string | null;
    complex?: null | {
        deepOpt?: string | null
    } 
}
type NullTest = WithoutNull<EditableModel>;

// @ts-expect-error: opt should remain optional
const a: NullTest = {}

const b : NullTest = {
    id: "hey"
}

const c : NullTest = {
    id: "hey",
    complex: {}
}

const d : NullTest = {
    id: "hey",
    opt: "ho",
    complex: {
        deepOpt: "hey"
    }
}
```

**Reply by ismail-codar** [reply]
Also it seems works.

**Comment by ismail-codar** [comment]
You are right. But this version can do it.  If I haven't made a mistake. I hope it happens this time!
```ts
type NonNullableObject<T> = {
  [K in keyof T]: T[K] extends undefined
    ? T[K]
    : NonNullableObject<NonNullable<T[K]>>;
};

```
![image](https://github.com/odata2ts/odata2ts/assets/282104/22308677-b280-4160-9ad4-bc726de043dd)

**Reply by texttechne** [reply, marked as answer]
Works and can be shortened - `T[K] extends undefined` never holds true. So:
```
type NonNullableObject<T> = {
  [K in keyof T]: NonNullableObject<NonNullable<T[K]>>;
};
```
See [Playground Example](https://www.typescriptlang.org/play?ssl=10&ssc=1&pln=12&pc=3#code/JYOwLgpgTgZghgYwgAgKIBNhjgIwDYQCyA9uhHsgN4CwAUMg8sOgFzIDOYUoA5gNx1GyYgAcwAfjaduIHsgA+yEAFc8eAfUYJiAWxEEAHpKWqKimpqENmbFTpzQANIKsMyEEQHkxx6bwUmai4MAL7IdCF0dGAAniIoAHLEIAmmuASeOABWEAhgADwAKgB8yAC8VC4A2gDSTCDIANYQMcQwyIUAumxJKWn4EJk5efm9qWrpEEW1ncXFGiEa0XGJpoUQnOXIY-0Z2bkFGFiTJGR4xVG0APRXyAACYOwAtBAG8XkvUFDEUGyiYBwABbEVToZBQCA6OCgYRiYDJOB4OjaECbOA9NYbAEVSiRWjI5KbHDIDFqdabHHBJisZAAIkBLVpEUuKM2CFJeHJ2MqlmpbHpjOcvP+-OBTNoeIJqIBNPGnKxWwsQhsdIZMVpQqEN3ujxebwOn2+v0CFHYwNByEBcAAbigHBAGjBgHhIBD0FSRSbmfjaKyASgOVzFVSVQL1ZqtLp9K82JQ+cgAIzIPGS32EgEwQMKym80NqjVU7UPZ6vd5gQ0-WymIEgvBgq225D2x3O10Qd287R6QxVoISlnp5A8LMUnnKmlhgudqM9seueMJiOudxeMSixlUlMD6WWkfcpWMPOCqld6MGWNU8dsReXxhF3Wlg3QI2903muuWm12iAO5BOl3QO2t5uD+q5gK+m4REAA)

Thanks, learned something here. I've thought your solution wouldn't work because the utility type `NonNullable` filters out `null` as well as `undefined`. Your solution works however, because **the keys themselves carry the information if a property is `undefined` or not** (via the question mark) and this information is left untouched by `[K in keyof T]` .
